### PR TITLE
JIT: Make BasicBlock::bbPrev and bbNext private

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Followup to #92908, and next step for #93020. CC @dotnet/jit-contrib.

### CODE REVIEW

**Comment by BruceForstall** [comment]
[No asm diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=427804&view=ms.vss-build-web.run-extensions-tab) but some TP diffs.

**Comment by BruceForstall** [comment]
Seems like we should work to avoid too many `GetBBNext()` calls in the code (maybe as a follow-up?)

E.g., instead of `(block->bbNext == nullptr)` => a new function `block->IsLast()`

`block->GetBBNext() == compiler->fgFirstColdBlock` => `block->IsLastHotBlock(compiler)` (true also for no cold section and `block->IsEnd()`?) Similarly, `block->IsFirstColdBlock(compiler)`?

File: src/coreclr/jit/block.h [hunk 1]
```diff
@@ -508,10 +508,49 @@ struct BasicBlock : private LIR::Range
...
+    void SetBBPrev(BasicBlock* prev)
+    {
+        bbPrev = prev;
+        if (prev)
+        {
+            prev->bbNext = this;
+        }
+    }
+
+    BasicBlock* GetBBNext() const
```
Comment (SingleAccretion) [inline src/coreclr/jit/block.h:548]: While we're renaming things, we could give these more "obvious" names, e. g. `GetNextBlock`.

**Comment by amanasifkhalid** [comment]
> [No asm diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=427804&view=ms.vss-build-web.run-extensions-tab) but some TP diffs (maybe because SetBBPrev() does more than what it replaces?)

We previously had BasicBlock::setNext() that updated bbNext, and also null-checked the new next block to see if it could update its bbPrev; most of the time, we'd just directly update bbNext, and would sparingly call setNext() if we also intended to update the next block's bbPrev. Would it make sense to not have SetBBNext() do this additional work, and keep setNext() around to do it? I think this would reduce/eliminate the TP diffs, but we'd lose the guarantee that bbNext and bbNext->bbPrev are consistent within SetBBNext().

> Seems like we should work to avoid too many GetBBNext() calls in the code (maybe as a follow-up?)

I'm happy to add helper methods for checking bbNext/bbPrev. This is for readability, since GetBBNext() should be getting inlined, right?

**Comment by amanasifkhalid** [comment]
> E.g., instead of `(block->bbNext == nullptr)` => a new function `block->IsLast()`, and `block->bbPrev == nullptr` => `block->IsFirst()`.

Would it be preferable to check against fgFirstBB and fgLastBB instead of adding new methods? Ditto checking if block == fgFirstColdBlock instead of a new method (though for checking if a block is the last hot block, I agree we'd need a new method to avoid calling GetBBNext()).

**Comment by BruceForstall** [comment]
> Would it make sense to not have SetBBNext() do this additional work, and keep setNext() around to do it?

No, it's probably best to leave it as you have it now.

> I'm happy to add helper methods for checking bbNext/bbPrev. This is for readability, since GetBBNext() should be getting inlined, right?

Yes.

> Would it be preferable to check against fgFirstBB and fgLastBB instead of adding new methods? Ditto checking if block == fgFirstColdBlock instead of a new method (though for checking if a block is the last hot block, I agree we'd need a new method to avoid calling GetBBNext()).

We don't always have a `Compiler*` available that can be used to get `fgFirstBB`/`fgLastBB`.

For `fgFirstColdBlock` we have no other way to check except using that variable, which then requires a `Compiler*`.

**Comment by AndyAyersMS** [comment]
I suspect over time we're going to see more and more cases where BB operations are contextual and need access to the compiler instance. At various times I've wondered if blocks should just point back to some parent object (either the compiler or some sort of "flow graph" object, but that seems wasteful. Given that blocks are all the same size we could perhaps play tricks to factor out common fields, if we say allocated them contiguously. For now, I think providing the context when needed via args is ok.

One issue with either a context arg or a parent object is that when we inline we do some funny things as there are two compiler objects around, and so it's possible to consult the wrong one....

**Comment by amanasifkhalid** [comment]
@BruceForstall @AndyAyersMS PTAL. CI is clean, but [small TP diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=428856&view=ms.vss-build-web.run-extensions-tab). I'm planning on making the jump target members private next; if we can assert the jump kind is valid every time we try to read the jump destination, then maybe we can remove code that sets the jump destination to null when changing the jump kind. This might make up for the TP regression in Release builds.

**Review by AndyAyersMS — APPROVED** [review]
This LGTM. Thanks!

File: src/coreclr/jit/block.cpp [hunk 2]
```diff
@@ -186,6 +186,21 @@ FlowEdge* Compiler::BlockPredsWithEH(BasicBlock* blk)
     return res;
 }
 
+//------------------------------------------------------------------------
+// IsLastHotBlock: see if this is the last block before the cold section
```
Comment (BruceForstall) [inline src/coreclr/jit/block.cpp:190]: Should we have a corresponding `IsFirstColdBlock(Compiler*)`?

File: src/coreclr/jit/block.h [hunk 3]
```diff
@@ -1435,10 +1476,10 @@ class BasicBlockIterator
     {
         assert(m_block != nullptr);
         // Check that we haven't been spliced out of the list.
-        assert((m_block->bbNext == nullptr) || (m_block->bbNext->bbPrev == m_block));
-        assert((m_block->bbPrev == nullptr) || (m_block->bbPrev->bbNext == m_block));
+        assert((m_block->IsLast()) || m_block->Next()->PrevIs(m_block));
+        assert((m_block->IsFirst()) || m_block->Prev()->NextIs(m_block));
```
Comment (BruceForstall) [inline src/coreclr/jit/block.h:1480]: With the function, we can remove an unnecessary set of parens
```suggestion
        assert(m_block->IsLast() || m_block->Next()->PrevIs(m_block));
        assert(m_block->IsFirst() || m_block->Prev()->NextIs(m_block));
```

File: src/coreclr/jit/codegencommon.cpp [hunk 4]
```diff
@@ -5210,7 +5210,7 @@ void CodeGen::genReserveEpilog(BasicBlock* block)
 
     assert(block != nullptr);
     const VARSET_TP& gcrefVarsArg(GetEmitter()->emitThisGCrefVars);
-    bool             last = (block->bbNext == nullptr);
+    bool             last = (block->IsLast());
```
Comment (BruceForstall) [inline src/coreclr/jit/codegencommon.cpp:5213]:
```suggestion
    bool             last = block->IsLast();
```
or just substitute it in the call below and eliminate the `last` var

File: src/coreclr/jit/codegencommon.cpp [hunk 5]
```diff
@@ -5257,7 +5257,7 @@ void CodeGen::genReserveFuncletEpilog(BasicBlock* block)
 
     JITDUMP("Reserving funclet epilog IG for block " FMT_BB "\n", block->bbNum);
 
-    bool last = (block->bbNext == nullptr);
+    bool last = (block->IsLast());
```
Comment (BruceForstall) [inline src/coreclr/jit/codegencommon.cpp:5260]: ditto

File: src/coreclr/jit/codegenlinear.cpp [hunk 6]
```diff
@@ -547,10 +546,10 @@ void CodeGen::genCodeForBBlist()
 
         /* Is this the last block, and are there any open scopes left ? */
 
-        bool isLastBlockProcessed = (block->bbNext == nullptr);
+        bool isLastBlockProcessed = (block->IsLast());
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenlinear.cpp:549]:
```suggestion
        bool isLastBlockProcessed = block->IsLast();
```

File: src/coreclr/jit/codegenlinear.cpp [hunk 7]
```diff
@@ -547,10 +546,10 @@ void CodeGen::genCodeForBBlist()
...
         if (block->isBBCallAlwaysPair())
         {
-            isLastBlockProcessed = (block->bbNext->bbNext == nullptr);
+            isLastBlockProcessed = (block->Next()->IsLast());
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenlinear.cpp:552]:
```suggestion
            isLastBlockProcessed = block->Next()->IsLast();
```

File: src/coreclr/jit/codegenlinear.cpp [hunk 8]
```diff
@@ -615,7 +614,7 @@ void CodeGen::genCodeForBBlist()
             // Note: we may be generating a few too many NOPs for the case of call preceding an epilog. Technically,
             // if the next block is a BBJ_RETURN, an epilog will be generated, but there may be some instructions
             // generated before the OS epilog starts, such as a GS cookie check.
-            if ((block->bbNext == nullptr) || !BasicBlock::sameEHRegion(block, block->bbNext))
+            if ((block->IsLast()) || !BasicBlock::sameEHRegion(block, block->Next()))
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenlinear.cpp:617]:
```suggestion
            if (block->IsLast() || !BasicBlock::sameEHRegion(block, block->Next()))
```

File: src/coreclr/jit/codegenlinear.cpp [hunk 9]
```diff
@@ -818,7 +817,7 @@ void CodeGen::genCodeForBBlist()
             GetEmitter()->emitLoopAlignment(DEBUG_ARG1(block->KindIs(BBJ_ALWAYS)));
         }
 
-        if ((block->bbNext != nullptr) && (block->bbNext->isLoopAlign()))
+        if (!block->IsLast() && (block->Next()->isLoopAlign()))
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenlinear.cpp:820]:
```suggestion
        if (!block->IsLast() && block->Next()->isLoopAlign())
```

File: src/coreclr/jit/compiler.cpp [hunk 10]
```diff
@@ -5291,11 +5291,11 @@ PhaseStatus Compiler::placeLoopAlignInstructions()
             }
         }
 
-        if ((block->bbNext != nullptr) && (block->bbNext->isLoopAlign()))
+        if (!block->IsLast() && (block->Next()->isLoopAlign()))
```
Comment (BruceForstall) [inline src/coreclr/jit/compiler.cpp:5294]:
```suggestion
        if (!block->IsLast() && block->Next()->isLoopAlign())
```

File: src/coreclr/jit/fgbasic.cpp [hunk 11]
```diff
@@ -5641,24 +5635,20 @@ void Compiler::fgMoveBlocksAfter(BasicBlock* bStart, BasicBlock* bEnd, BasicBloc
     {
         printf("Relocated block%s [" FMT_BB ".." FMT_BB "] inserted after " FMT_BB "%s\n", (bStart == bEnd) ? "" : "s",
                bStart->bbNum, bEnd->bbNum, insertAfterBlk->bbNum,
-               (insertAfterBlk->bbNext == nullptr) ? " at the end of method" : "");
+               (insertAfterBlk->IsLast()) ? " at the end of method" : "");
```
Comment (BruceForstall) [inline src/coreclr/jit/fgbasic.cpp:5638]:
```suggestion
               insertAfterBlk->IsLast() ? " at the end of method" : "");
```

**Review by BruceForstall — COMMENTED** [review]
Found a few nits that could be updated. Can't say I looked at every case.

File: src/coreclr/jit/block.cpp (hunk 2 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/block.cpp:190]: I can add that in for consistency.

File: src/coreclr/jit/block.h (hunk 3 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/block.h:1480]: Thanks for catching all these nits (sorry, these kinds of PRs seem especially tedious to review).

**Comment by amanasifkhalid** [comment]
@BruceForstall I think I fixed all the style issues introduced (plus a couple extra I found along the way).

**Review by BruceForstall — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**